plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt

```diff
@@ -98,9 +101,12 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
 
         // Note: it is important that `body` is only set after `embedUserFunction` is complete, as we need to
         // place the embedding in the map before processing the body.
-        embedUserFunction(declaration.symbol, signature).apply {
+        val userFunc = embedUserFunction(declaration.symbol, signature).apply {
             body = stmtCtx.convertMethodWithBody(declaration, signature, returnTarget)
         }
+
+        // For now this will just traverse the ExpEmbedding Tree and check whether the asserts are true
+        purityChecker.checkIsPure(userFunc)
```

Comment:
It doesn't make sense to check the purity of the function here, since we don't know that this function is pure.  Rather, what you want is to check the correctness of the ExpEmbedding.